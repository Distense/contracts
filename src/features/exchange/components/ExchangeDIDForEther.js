import React, { Component } from 'react'
import { BigNumber } from 'bignumber.js'
import { Button, Input, Form, Grid, List, Message } from 'semantic-ui-react'

import { exchangeDIDForEther } from '../actions'

export default class ExchangeDIDForEther extends Component {
  constructor(props) {
    super(props)
    this.state = {
      numDIDToExchange: '',
      numEtherUserWillReceive: ''
    }

    this.onChangeNumDID = this.onChangeNumDID.bind(this)
  }

  onChangeNumDID = ({ target: { value } }) => {
    const numDIDUserMayExchange = this.props.numDIDUserMayExchange
    const didPerEtherExchangeRate = this.props.didPerEtherExchangeRate

    value = +value
    if (value > numDIDUserMayExchange) value = numDIDUserMayExchange

    const numEtherUserWillReceive = new BigNumber(value)
      .div(didPerEtherExchangeRate)
      .dp(4)
      .toString()

    //  If user has entered a value and value is still within bounds of how many DID they can exchange into ether
    if (value > 0 && value <= numDIDUserMayExchange) {
      this.setState({
        numEtherUserWillReceive,
        numDIDToExchange: value
      })
    } else if (value > numDIDUserMayExchange) {
      this.setState({
        numDIDToExchange: numDIDUserMayExchange,
        numEtherUserWillReceive
      })
    } else {
      this.setState({
        numDIDToExchange: '',
        numEtherUserWillReceive
      })
    }
  }

  onSubmitExchangeDIDForEther = async e => {
    e.preventDefault()

    const { numDIDToExchange } = this.state
    exchangeDIDForEther({ numDIDToExchange })
  }

  render() {
    const { numDIDToExchange, numEtherUserWillReceive } = this.state

    const { numDIDUserMayExchange, numDIDOwned } = this.props

    return (
      <Grid.Column width={8}>
        <Form onSubmit={this.onSubmitExchangeDIDForEther}>
          <Form.Field>
            <Input
              onChange={this.onChangeNumDID}
              placeholder={`number of DID you want to exchange into ether`}
              value={numDIDToExchange}
            />
          </Form.Field>
          <Button size="medium" color="green" type="submit">
            Exchange {numDIDToExchange} DID for {numEtherUserWillReceive} ether
          </Button>
        </Form>
        <Message>
          <Message.Header>Exchange DID For ETH</Message.Header>
          <List bulleted>
            {numDIDOwned ? (
              <List.Item>
                DID exchangeable by you: {numDIDUserMayExchange} DID
              </List.Item>
            ) : (
              <List.Item>
                You don't own any DID in the selected Ethereum account so you
                have no DID to exchange
              </List.Item>
            )}
            <List.Item>
              You are relinquishing Distense governance/voting rights by
              exchanging your DID for ether.
            </List.Item>
            <List.Item>
              This is a permanent, non-reversible, undoable, final action. There
              is no one who can undo this.
            </List.Item>
            <List.Item>
              If this says you can't invest, try reloading the page or ensure
              MetaMask is unlocked
            </List.Item>
          </List>
        </Message>
      </Grid.Column>
    )
  }
}
