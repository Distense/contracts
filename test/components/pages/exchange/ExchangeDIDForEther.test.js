import React from 'react'
import { shallow } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import configureMockStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import ExchangeDIDForEther from '../../../../src/pages/Exchange/ExchangeDIDForEther'

function setup(
  numDIDUserMayExchange = null,
  numDIDOwned = null,
  didPerEtherExchangeRate
) {
  const props = {
    numDIDUserMayExchange,
    numDIDOwned,
    didPerEtherExchangeRate,
    exchangeDIDForEther: jest.fn()
  }
  let wrapper = shallow(<ExchangeDIDForEther {...props} />)

  return { wrapper, props }
}

describe('<ExchangeDIDForEther /> page component', () => {
  it('should contain some basic subcomponents', () => {
    const { wrapper } = setup()
    expect(wrapper.find('Button').length).toEqual(1)
    expect(wrapper.find('GridColumn').length).toEqual(1)
    expect(wrapper.find('Head').length).toEqual(0)
    expect(wrapper.find('Form').length).toEqual(1)
    expect(wrapper.find('FormField').length).toEqual(1)
    expect(wrapper.find('List').length).toEqual(1)
    expect(wrapper.find('Message').length).toEqual(1)
    expect(wrapper.find('MessageHeader').length).toEqual(1)
  })

  it('should set the initial state correctly', () => {
    const { wrapper } = setup()
    expect(wrapper.state('numDIDToExchange')).toEqual('')
    expect(wrapper.state('numEtherUserWillReceive')).toEqual('')
  })

  //  TODO this throws because the button now has three children and not one node
  // it('should contain one button exchange DID for ether)', () => {
  //   const { wrapper } = setup(1000, 1000, 1000)
  //   const mockedEvent = { target: { value: '100' } }
  //   wrapper.instance().onChangeNumDID(mockedEvent)
  //   expect(
  //     wrapper
  //       .find('Button')
  //       .children()
  //       .text()
  //   ).toEqual('Exchange 100 DID for 0.1 ether')
  // })

  it('should correctly determine the number of ether the user will receive', () => {
    const { wrapper } = setup(500, 1000, 1000)
    const mockedEvent = { target: { value: 300 } }
    wrapper.instance().onChangeNumDID(mockedEvent)
    expect(wrapper.state('numDIDToExchange')).toEqual(300)
    expect(wrapper.state('numEtherUserWillReceive')).toEqual('0.3')
  })

  it('should correctly determine the number of ether the user will receive', () => {
    const { wrapper } = setup(500, 1000, 1000)
    const mockedEvent = { target: { value: 324 } }
    wrapper.instance().onChangeNumDID(mockedEvent)
    expect(wrapper.state('numDIDToExchange')).toEqual(324)
    expect(wrapper.state('numEtherUserWillReceive')).toEqual('0.324')
  })

  it('should limit the value of numDIDToExchange to numDIDUserMayExchange', () => {
    const { wrapper } = setup(500, 1000, 1000)
    const mockedEvent = { target: { value: 501 } }
    wrapper.instance().onChangeNumDID(mockedEvent)
    expect(wrapper.state('numDIDToExchange')).toEqual(500)
    expect(wrapper.state('numEtherUserWillReceive')).toEqual('0.5')
  })

  it('should reset the value of numDIDToExchange to an empty string', () => {
    const { wrapper } = setup(500, 1000, 1000)
    const mockedEvent = { target: { value: '' } }
    wrapper.instance().onChangeNumDID(mockedEvent)
    expect(wrapper.state('numDIDToExchange')).toEqual('')
    expect(wrapper.state('numEtherUserWillReceive')).toEqual('0')
  })

  test('snapshot', () => {
    const { props } = setup('executive')
    const mockstore = configureMockStore()
    const store = mockstore({})
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <ExchangeDIDForEther {...props} />
          </MemoryRouter>
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
