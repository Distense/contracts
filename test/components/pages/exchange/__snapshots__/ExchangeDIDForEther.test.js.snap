// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExchangeDIDForEther /> page component snapshot 1`] = `
<div
  className="eight wide column"
>
  <form
    action={undefined}
    className="ui form"
    onSubmit={[Function]}
  >
    <div
      className="field"
    >
      <div
        className="ui input"
      >
        <input
          disabled={undefined}
          onChange={[Function]}
          placeholder="number of DID you want to exchange into ether"
          tabIndex={undefined}
          type="text"
          value=""
        />
      </div>
    </div>
    <button
      className="ui green medium button"
      disabled={undefined}
      onClick={[Function]}
      role="button"
      tabIndex={undefined}
      type="submit"
    >
      Exchange 
      
       DID for 
      
       ether
    </button>
  </form>
  <div
    className="ui message"
  >
    <div
      className="header"
    >
      Exchange DID For ETH
    </div>
    <div
      className="ui bulleted list"
      role="list"
    >
      <div
        className="item"
        data-value={undefined}
        onClick={[Function]}
        role="listitem"
      >
        You don't own any DID in the selected Ethereum account so you have no DID to exchange
      </div>
      <div
        className="item"
        data-value={undefined}
        onClick={[Function]}
        role="listitem"
      >
        You are relinquishing Distense governance/voting rights by exchanging your DID for ether.
      </div>
      <div
        className="item"
        data-value={undefined}
        onClick={[Function]}
        role="listitem"
      >
        This is a permanent, non-reversible, undoable, final action. There is no one who can undo this.
      </div>
      <div
        className="item"
        data-value={undefined}
        onClick={[Function]}
        role="listitem"
      >
        If this says you can't invest, try reloading the page or ensure MetaMask is unlocked
      </div>
    </div>
  </div>
</div>
`;
